### ISSUE TITLE
Compile error on Ruby 2.3.0

### ISSUE THREAD

**Issue by mattrei** [wedesoft/hornetseye-v4l2#12]
Hi,

I have following version: `ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-linux]`
However I get the following compile error

```
In file included from /usr/include/ruby-2.3.0/ruby.h:33:0,
                 from ext/rubyinc.hh:25,
                 from ext/v4l2select.cc:16:
ext/v4l2select.cc: In member function ‘unsigned int V4L2Select::coding()’:
/usr/include/ruby-2.3.0/ruby/ruby.h:882:52: error: ‘__builtin_choose_expr’ was not declared in this scope
      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
                                                    ^
/usr/include/ruby-2.3.0/ruby/ruby.h:686:41: note: in definition of macro ‘RB_NUM2INT’
 #define RB_NUM2INT(x) rb_num2int_inline(x)
                                         ^
ext/v4l2select.cc:54:10: note: in expansion of macro ‘NUM2INT’
   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 0 ]);
          ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:33: note: in expansion of macro ‘RARRAY_CONST_PTR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                 ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:50: note: in expansion of macro ‘RB_OBJ_WB_UNPROTECT_FOR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                                  ^
ext/v4l2select.cc:54:18: note: in expansion of macro ‘RARRAY_PTR’
   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 0 ]);
                  ^
ext/v4l2select.cc: In member function ‘unsigned int V4L2Select::width()’:
/usr/include/ruby-2.3.0/ruby/ruby.h:882:52: error: ‘__builtin_choose_expr’ was not declared in this scope
      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
                                                    ^
/usr/include/ruby-2.3.0/ruby/ruby.h:686:41: note: in definition of macro ‘RB_NUM2INT’
 #define RB_NUM2INT(x) rb_num2int_inline(x)
                                         ^
ext/v4l2select.cc:59:10: note: in expansion of macro ‘NUM2INT’
   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 1 ]);
          ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:33: note: in expansion of macro ‘RARRAY_CONST_PTR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                 ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:50: note: in expansion of macro ‘RB_OBJ_WB_UNPROTECT_FOR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                                  ^
ext/v4l2select.cc:59:18: note: in expansion of macro ‘RARRAY_PTR’
   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 1 ]);
                  ^
ext/v4l2select.cc: In member function ‘unsigned int V4L2Select::height()’:
/usr/include/ruby-2.3.0/ruby/ruby.h:882:52: error: ‘__builtin_choose_expr’ was not declared in this scope
      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
                                                    ^
/usr/include/ruby-2.3.0/ruby/ruby.h:686:41: note: in definition of macro ‘RB_NUM2INT’
 #define RB_NUM2INT(x) rb_num2int_inline(x)
                                         ^
ext/v4l2select.cc:64:10: note: in expansion of macro ‘NUM2INT’
   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 2 ]);
          ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:33: note: in expansion of macro ‘RARRAY_CONST_PTR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                 ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:50: note: in expansion of macro ‘RB_OBJ_WB_UNPROTECT_FOR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                                  ^
ext/v4l2select.cc:64:18: note: in expansion of macro ‘RARRAY_PTR’
   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 2 ]);
                  ^
rake aborted!
Command failed with status (1): [g++ -DNDEBUG -D_FORTIFY_SOURCE=2   -march=...]
/home/mat/.gem/ruby/2.3.0/gems/hornetseye-v4l2-1.0.6/Rakefile:78:in `block in <top (required)>'
Tasks: TOP => default => all => ext/hornetseye_v4l2.so => ext/v4l2select.o
(See full trace by running task with --trace)
```

With 2.2.x everything compiles smoothly.

Thx,
Best

**Comment by wedesoft** [comment]
On January 6, 2016 5:42:13 PM GMT, Matthias Treitler <notifications@github.com> wrote:
>Hi,
>
>I have following version: `ruby 230p0 (2015-12-25 revision 53290)
>[x86_64-linux]`
>However I get the following compile error
>```
>In file included from /usr/include/ruby-230/rubyh:33:0,
>                 from ext/rubyinchh:25,
>                 from ext/v4l2selectcc:16:
>ext/v4l2selectcc: In member function ‘unsigned int
>V4L2Select::coding()’:
>/usr/include/ruby-230/ruby/rubyh:882:52: error: ‘__builtin_choose_expr’
>was not declared in this scope
>      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
>                                                    ^
>/usr/include/ruby-230/ruby/rubyh:686:41: note: in definition of macro
>‘RB_NUM2INT’
> #define RB_NUM2INT(x) rb_num2int_inline(x)
>                                         ^
>ext/v4l2selectcc:54:10: note: in expansion of macro ‘NUM2INT’
>   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 0 ]);
>          ^
>/usr/include/ruby-230/ruby/rubyh:1059:33: note: in expansion of macro
>‘RARRAY_CONST_PTR’
>#define RARRAY_PTR(a) ((VALUE
>*)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
>                                 ^
>/usr/include/ruby-230/ruby/rubyh:1059:50: note: in expansion of macro
>‘RB_OBJ_WB_UNPROTECT_FOR’
>#define RARRAY_PTR(a) ((VALUE
>*)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
>                                                  ^
>ext/v4l2selectcc:54:18: note: in expansion of macro ‘RARRAY_PTR’
>   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 0 ]);
>                  ^
>ext/v4l2selectcc: In member function ‘unsigned int
>V4L2Select::width()’:
>/usr/include/ruby-230/ruby/rubyh:882:52: error: ‘__builtin_choose_expr’
>was not declared in this scope
>      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
>                                                    ^
>/usr/include/ruby-230/ruby/rubyh:686:41: note: in definition of macro
>‘RB_NUM2INT’
> #define RB_NUM2INT(x) rb_num2int_inline(x)
>                                         ^
>ext/v4l2selectcc:59:10: note: in expansion of macro ‘NUM2INT’
>   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 1 ]);
>          ^
>/usr/include/ruby-230/ruby/rubyh:1059:33: note: in expansion of macro
>‘RARRAY_CONST_PTR’
>#define RARRAY_PTR(a) ((VALUE
>*)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
>                                 ^
>/usr/include/ruby-230/ruby/rubyh:1059:50: note: in expansion of macro
>‘RB_OBJ_WB_UNPROTECT_FOR’
>#define RARRAY_PTR(a) ((VALUE
>*)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
>                                                  ^
>ext/v4l2selectcc:59:18: note: in expansion of macro ‘RARRAY_PTR’
>   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 1 ]);
>                  ^
>ext/v4l2selectcc: In member function ‘unsigned int
>V4L2Select::height()’:
>/usr/include/ruby-230/ruby/rubyh:882:52: error: ‘__builtin_choose_expr’
>was not declared in this scope
>      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
>                                                    ^
>/usr/include/ruby-230/ruby/rubyh:686:41: note: in definition of macro
>‘RB_NUM2INT’
> #define RB_NUM2INT(x) rb_num2int_inline(x)
>                                         ^
>ext/v4l2selectcc:64:10: note: in expansion of macro ‘NUM2INT’
>   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 2 ]);
>          ^
>/usr/include/ruby-230/ruby/rubyh:1059:33: note: in expansion of macro
>‘RARRAY_CONST_PTR’
>#define RARRAY_PTR(a) ((VALUE
>*)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
>                                 ^
>/usr/include/ruby-230/ruby/rubyh:1059:50: note: in expansion of macro
>‘RB_OBJ_WB_UNPROTECT_FOR’
>#define RARRAY_PTR(a) ((VALUE
>*)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
>                                                  ^
>ext/v4l2selectcc:64:18: note: in expansion of macro ‘RARRAY_PTR’
>   return NUM2INT(RARRAY_PTR(m_rbSelection)[ 2 ]);
>                  ^
>rake aborted!
>Command failed with status (1): [g++ -DNDEBUG -D_FORTIFY_SOURCE=2  
>-march=]
>/home/mat/gem/ruby/230/gems/hornetseye-v4l2-106/Rakefile:78:in `block
>in <top (required)>'
>Tasks: TOP => default => all => ext/hornetseye_v4l2so =>
>ext/v4l2selecto
>(See full trace by running task with --trace)
>```
>
>With 22x everything compiles smoothly
>
>Thx,
>Best
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/wedesoft/hornetseye-v4l2/issues/12

Hi,
Looks like this is because of an API change:
https://github.com/robinst/taglib-ruby/pull/68

Currently busy but hope to get around to fix it this weekend.
-- 
Jan Wedekind
http://www.wedesoft.de/

**Comment by wedesoft** [comment]
Looks like the Ruby 2.3.0 native API now relies on [__builtin_choose_expr](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html) which is available in C but not in C++. I haven't managed to resolve this yet.

**Comment by mattrei** [comment]
Oh I'm sorry for that. 
My project luckily does not specify any specific Ruby version so I can also use 2.2. However C ABIs are compatible to C++, not?

**Comment by wedesoft** [comment]
Looks like replacing RARRAY_PTR with the alternatives RARRAY_CONST_PTR, rb_ary_entry, and rb_ary_store fixes the problem. I released [hornetseye-v4l2 version 1.0.7](https://rubygems.org/gems/hornetseye-v4l2/versions/1.0.7).

**Comment by wedesoft** [comment]
I'll close this issue. Let me know if this does not fix compatibility with Ruby 2.3.0.
